File: engines/corese/src/main/kotlin/BenchmarkRunner.kt
package es.weso.rudof

import java.io.File
import java.util.concurrent.Callable
import java.util.concurrent.ExecutionException
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException
import kotlin.time.measureTimedValue

class BenchmarkRunner<R>(
    private val engine: ValidationEngine<R>,
    private val args: Args,
) {
    private data class RunOutcome(val loadTimeMs: Double, val validationTimeMs: Double, val report: String)

    fun run(): ResultState {
        val results = ResultState(args.minValidIter)

        repeat(args.warmUp + args.runs) { idx ->
            val (iterationResult, report) = runIteration(idx)

            if (idx >= args.warmUp) {
                results.addResult(iterationResult.copy(iteration = iterationResult.iteration - args.warmUp))
                report?.let { File(args.reportPath).writeText(it) }
            }
            if (idx == args.warmUp - 1) {
                println("[${engine.name}] Warm-up complete")
            }
        }

        return results
    }

    private fun runIteration(idx: Int): Pair<IterationResults, String?> {
        val executor = Executors.newSingleThreadExecutor { r ->
            Thread(r, "${engine.name}-run-$idx").apply { isDaemon = true }
        }

        val future = executor.submit(Callable {
            System.gc()
            val load = measureTimedValue { engine.loadData(args.dataPath, args.dataFormat, args.shapesPath, args.shapesFormat) }
            System.gc()
            val validation = measureTimedValue { engine.validate() }
            val report = engine.generateReport(validation.value)
            RunOutcome(
                loadTimeMs = load.duration.inWholeMicroseconds / 1000.0,
                validationTimeMs = validation.duration.inWholeMicroseconds / 1000.0,
                report = report,
            )
        })

        return try {
            val outcome = future.get(args.timeout.toLong(), TimeUnit.SECONDS)
            val result = IterationResults(idx, outcome.loadTimeMs, outcome.validationTimeMs, IterationResultStatus.OK)
            result to outcome.report
        } catch (_: TimeoutException) {
            println("[${engine.name}] Run $idx exceeded timeout of ${args.timeout}s")
            future.cancel(true)
            IterationResults(idx, null, null, IterationResultStatus.DNF) to null
        } catch (e: ExecutionException) {
            val cause = e.cause
            val status = if (cause is OutOfMemoryError) IterationResultStatus.OOM else IterationResultStatus.ERROR
            println("[${engine.name}] Run $idx failed: ${cause?.message}")
            IterationResults(idx, null, null, status) to null
        } finally {
            executor.shutdownNow()
        }
    }
}

File: engines/maplib/benchmark_runner.py
import gc
import time
from concurrent.futures import ThreadPoolExecutor, TimeoutError as FutureTimeoutError
from dataclasses import replace

from args import Args
from results import IterationResults, IterationResultStatus, ResultState
from validation_engine import ValidationEngine

class BenchmarkRunner:
    def __init__(self, engine: ValidationEngine, args: Args):
        self._engine = engine
        self._args = args

    def run(self) -> ResultState:
        results = ResultState(self._args.min_valid_iter)

        for idx in range(self._args.warm_up + self._args.runs):
            iteration_result, report = self._run_iteration(idx)

            if idx >= self._args.warm_up:
                results.add_result(replace(iteration_result, iteration=iteration_result.iteration - self._args.warm_up))
                if report is not None:
                    with open(self._args.report_path, "w", encoding="utf-8") as f:
                        f.write(report)

            if idx == self._args.warm_up - 1:
                print(f"[{self._engine.name}] Warm-up complete")

        return results

    def _run_iteration(self, idx: int) -> tuple[IterationResults, str | None]:
        executor = ThreadPoolExecutor(max_workers=1, thread_name_prefix=f"{self._engine.name}-run-{idx}")
        future = executor.submit(self._iteration_body)

        try:
            load_ms, validation_ms, report = future.result(timeout=self._args.timeout)
            return IterationResults(idx, load_ms, validation_ms, IterationResultStatus.OK), report
        except FutureTimeoutError:
            print(f"[{self._engine.name}] Run {idx} exceeded timeout of {self._args.timeout}s")
            return IterationResults(idx, None, None, IterationResultStatus.DNF), None
        except MemoryError as e:
            print(f"[{self._engine.name}] Run {idx} failed: {e}")
            return IterationResults(idx, None, None, IterationResultStatus.OOM), None
        except Exception as e:
            print(f"[{self._engine.name}] Run {idx} failed: {e}")
            return IterationResults(idx, None, None, IterationResultStatus.ERROR), None
        finally:
            executor.shutdown(wait=False, cancel_futures=True)

    def _iteration_body(self) -> tuple[float, float, str]:
        gc.collect()
        gc.disable()
        load_start = time.perf_counter_ns()
        self._engine.load_data(
            self._args.data_path,
            self._args.data_format,
            self._args.shapes_path,
            self._args.shapes_format,
        )
        load_ms = (time.perf_counter_ns() - load_start) / 1_000_000.0
        gc.enable()

        gc.collect()
        gc.disable()
        validation_start = time.perf_counter_ns()
        result = self._engine.validate()
        validation_ms = (time.perf_counter_ns() - validation_start) / 1_000_000.0
        gc.enable()

        report = self._engine.generate_report(result)
        return load_ms, validation_ms, report
